### PR TITLE
socket: implement context support

### PR DESCRIPTION
**Pull request by dwlnetnl**
Pull request for #6.

### CODE REVIEW

**Review by mdlayher — COMMENTED** [review]
Thanks for doing this, it appears you have a bit more BSD sockets internals knowledge than I do.

Since this package is a building block of best practices and is not at v1.0.0, I would not object to just adding the context as the first parameter to all the special system calls but keeping the same names.

We still want to implement net.Conn and such so we have to keep Read/Write/Close and the like as-is though.

But internally if we can assume context support everywhere then we can remove a lot of the conditionals. For example, if we pass a context.Background then you can just check ctx.Err no matter what because it'll always be nil I believe.

How does that sound?

File: conn.go [hunk 1]
```diff
@@ -344,6 +422,11 @@ func (c *Conn) Bind(sa unix.Sockaddr) error {
 // connected to a remote peer, Connect calls getpeername(2) and returns the
 // unix.Sockaddr from that call.
 func (c *Conn) Connect(sa unix.Sockaddr) (unix.Sockaddr, error) {
+	return c.ConnectContext(noContext, sa)
+}
+
+// ConnectContext is like Connect but takes a context.
+func (c *Conn) ConnectContext(ctx context.Context, sa unix.Sockaddr) (unix.Sockaddr, error) {
```
Comment (mdlayher) [inline conn.go:429]: As an example of what I mean, we can just make Connect:

```go
func (c *Conn) Connect(ctx context.Context, sa unix.Sockaddr) (unix.Sockaddr, error) {
```

And drop the suffix entirely.

Comment (dwlnetnl) [reply conn.go:429]: Haha, I am not sure if I have more knowledge of the BSD sockets, I wouldn't say I had any when I started this! Mostly a lot spelunking through the manual pages and net, os and internal/poll package.

I like to just put a context in as a first parameter, out of politeness and backward compatibility I opted to duplicate most methods. So remove the duplication. What about the comment about the order of parameters in the Sendto method? Adding a context parameter is probably a good moment to change the order as well as you have to change the call sides anyway.

Comment (dwlnetnl) [reply conn.go:429]: How do you want to go about Read and Write as *os.File provides a internal/poll implementation for them and if no context is required, I don't see any reason to use custom code. There will be an if though if a context is passed as the first parameter.

Comment (mdlayher) [reply conn.go:429]: > I like to just put a context in as a first parameter, out of politeness and backward compatibility I opted to duplicate most methods. So remove the duplication.

Yep I totally understand, but in this case I'd rather just break the few users of this package up front so we can fix it before a v1.0.0. I can tag a new minor version after we merge this.

> What about the comment about the order of parameters in the Sendto method? Adding a context parameter is probably a good moment to change the order as well as you have to change the call sides anyway.

Yes please, thanks. I forgot about that TODO.

> How do you want to go about Read and Write as *os.File provides a internal/poll implementation for them and if no context is required, I don't see any reason to use custom code. There will be an if though if a context is passed as the first parameter.

SGTM, if the sentinel noContext makes more sense then it seems fine to retain it in these cases. But I suspect for the majority of functions we could just pass a background context to keep it simple.

**Comment by dwlnetnl** [comment]
@mdlayher do you have an idea why the TestLinuxConnBuffers test fails?

**Comment by mdlayher** [comment]
No idea actually, but it seems to work fine on main; just not your branch. I wonder what invariant has changed.

**Comment by dwlnetnl** [comment]
It doesn't seem to be near any context code. That baffles me a bit.

**Comment by dwlnetnl** [comment]
I found the issue, it was a bit of silly mistake. An error wasn't copied outside of RawConn closure.

**Comment by dwlnetnl** [comment]
It’s ready from my end to be merged in.

**Comment by mdlayher** [comment]
Thank you! I want to play with this a bit in a couple of my libraries to verify the API changes, and I may tweak the docs a bit. But I appreciate your time and effort on this.

**Comment by dwlnetnl** [comment]
Feel free! Absolutely check it out first.
I’m glad that you’re open to this addition!

**Comment by mdlayher** [comment]
See #8, I'll end up squashing and merging there! Thank you.

**Pull request closed without merging**